src/prefect/core/task.py

```diff
@@ -368,6 +373,28 @@ def __init__(
             raise ValueError(
                 "A datetime.timedelta `retry_delay` must be provided if max_retries > 0"
             )
+        if retry_on and not max_retries > 0:
+            raise ValueError(
+                "A numer of `max_retries` must be provided if `retry_on` is set."
+            )
+        if retry_on:
+            try:
+                retry_on = set(retry_on)
+            except TypeError as exc:
+                raise TypeError(
+                    "Unexpected type for `retry_on`. "
+                    "Expected an iterable of exception types."
+                ) from exc
+            for v in retry_on:
+                if not isinstance(v, type):
```

Comment:
Would the next if statement already catch this?
```
if not issubclass(v, Exception):  
```

Also, I'm confused about checking against `type`

Reply:
Okay TIL you can isinstance against type, but still wondering what this catches that `issubclass` doesn't
```
>>> isinstance(ValueError, type)
True
>>> isinstance(3, type)
False
>>> isinstance(ValueError, Exception)
False
>>> issubclass(ValueError, Exception)
True
```

Reply:
We'll throw an exception if the first argument to `issubclass` is not a type so we need to check that first

```
In [2]: issubclass(ValueError(), Exception)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [2], in <cell line: 1>()
----> 1 issubclass(ValueError(), Exception)

TypeError: issubclass() arg 1 must be a class
```